src/EditorFeatures/Core/Implementation/AutomaticCompletion/BraceCompletionSessionProvider.BraceCompletionSession.cs

```diff
@@ -100,7 +101,14 @@ private void Start(CancellationToken cancellationToken)
 
                 OpeningPoint = SubjectBuffer.CurrentSnapshot.CreateTrackingPoint(openingSnapshotPoint, PointTrackingMode.Positive);
 
-                var braceResult = _session.GetBraceCompletion(this, cancellationToken);
+                var context = GetBraceCompletionContext();
+                if (context == null)
+                {
+                    EndSession();
+                    return;
+                }
+
+                var braceResult = _service.GetBraceCompletionAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
                 if (braceResult == null)
                 {
                     EndSession();
```

Comment:
so ths pattern of having multiple locations need to 'EndSession' is error prone.  i.e. you could have easily added:

```
                if (context == null)
                {
                    return;
                }
```

and we likely wouldn't have noticed a bug. 

a nicer pattern i've found is more like:

```
if (!DoTheWork(args))
    EndSession();

//...

bool DoTheWork(... parameters ....)
{
    if (context == null)
        return false;
}

```

The benefit here is that the callee is forced to return a value, so they have to think about success/failure.  and tehn all failure is handle din a uniform fashion.



And